grr/server/grr_response_server/gui/ui/lib/api_translation/primitive.ts

```diff
@@ -71,3 +71,72 @@ export function createUnknownObject(anyObject?: AnyObject): unknown|undefined {
   delete result['@type'];
   return result;
 }
+
+/**
+ * Decodes a encoded base64 string into a byte array.
+ * On error, returns an empty array.
```

Comment:
Why is an empty array returned on error? This sort of masking makes actual issues harder to track and should only be used if reasons are well-documented.

Reply:
I don't see this always as an issue, because also throwing errors can result in unexpected behaviours (remember ClientLabel?). One other way is to return undefined on error, as this will let the user expect a different output on error.

Reply:
The unexpected behavior with ClientLabels was that ClientLabels did not show up at all. This was caused by an inconsistency in our API: Some API endpoints populated the `owner` field and some did not. The code that converts from Protobuf-JSON (ApiClientLabel) to the Frontend's internal data structure (ClientLabel) rightfully threw an Exception, which led us to discover this inconsistency in the first place.

`decodeBase64` is a library function that could be re-used in many places. If you return a special value on error (empty array or undefined), all users of your library function need to check if the returned value is the special error value on every use. If they don't, their code will likely fail (e.g. the IP address parser expects the array to have a certain length and fails) with a hard-to-debug Exception ("undefined is not a function" anyone?). Then, as a developer, it can be really hard to trace the root cause.

Returning special values that need to be checked for has similarities to golang's error handling and Java's checked exceptions. It incurs a lot of boilerplate code. Errors during parsing of base64 are not expected, but should only occur in *exceptional* circumstances, thus throwing an Exception is the right thing to do.